### PR TITLE
New: no-double-assignments (fixes #3481, cleaner #3482)

### LINKED ISSUE #3481: Rule proposal: no-double-assignments
Overwriting a variable twice in a row is most likely an error:

```js
cursor_x = 53;
cursor_x = 62;
```

The code above most likely is a result of a copy and paste error and the correct code should be

```js
cursor_x = 53;
cursor_y = 62;
```

See for a comparison [rule V519](http://www.viva64.com/en/d/0108/) in PVS Studio linter for C.

It seems to me to be a fairly useful error detection mechanism, suitable for core.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26057863-rule-proposal-no-double-assignments?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by denis-sokolov**

### CODE REVIEW

**Comment by michaelficarra** [comment]
I don't like this, especially its limitation that it can only detect adjacent statements. I would solve this with the scope analysis. You can look at all of the write references and determine if any of them do not include a read reference between them. Still, this is going to be difficult for cases like

```js
a = 0;
a = f();
function f(){ return a + 1; }
```

without a lot more code. Also, I think these limitations should be documented.

**Comment by denis-sokolov** [comment]
It’s not meant to solve the problem ideally. 20% effort for 80% of results is good enough. False negatives is not a problem.

False positives are a problem, but the example above is very contrived. I’d love to see a slightly more realistic false positive.

File: lib/rules/no-double-assignments.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+//------------------------------------------------------------------------------
+
+/**
+ * Return a node previous to the current
+ * @param {RuleContext} context The rule context
+ * @param {ASTNode} node The current Node
+ * @returns {ASTNode} The previous node
+ */
+function getPrevNode(context, node) {
+    var prevToken = context.getTokenBefore(node);
```
Comment (btmills) [inline lib/rules/no-double-assignments.js]: These methods have been moved from [`context`](http://eslint.org/docs/developer-guide/working-with-rules#the-context-object) to [`SourceCode`](http://eslint.org/docs/developer-guide/working-with-rules#contextgetsourcecode). You can get `SourceCode` from `context.getSourceCode()` once and pass it in whenever you need it.

File: tests/lib/rules/no-double-assignments.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        "foo.bar.x = 1; foo.quux.x = 2",
+
+        "foo = []; foo = add(foo);",
+        "foo = []; foo = add(foo.concat(bar));",
+        "foo = ''; foo = bar + foo",
+
+        // Whitespace tests
+        "foo = 5;bar = 5;",
+        "foo=5;bar=5;",
+        "   foo = 5   ; "
```
Comment (btmills) [inline tests/lib/rules/no-double-assignments.js]: Can you add a whitespace test that spans multiple lines as well?

File: tests/lib/rules/no-double-assignments.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    ],
+    invalid: [
+        { code: "foo = 5; foo = 6;", errors: defaultErrors },
```
Comment (btmills) [inline tests/lib/rules/no-double-assignments.js]: Can this be made to cover cases like `var foo = 5; foo = 6;`?

File: lib/rules/no-double-assignments.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
+/**
+ * @fileoverview Rule to catch accidental variable overwriting
+ * @author Denis Sokolov
+ * @copyright 2015 Denis Sokolov. All rights reserved.
+ */
```
Comment (nzakas) [inline lib/rules/no-double-assignments.js]: Please add license line. See: http://eslint.org/docs/developer-guide/code-conventions#file-format

**Comment by nzakas** [comment]
Agree with @michaelficarra in that using the escope information is going to be a lot more reliable and require a lot less code to accomplish the same thing.

**Comment by denis-sokolov** [comment]
I have updated the code according to the comments.

I originally hoped that this approach is a simple and cheap way to catch some user errors.
It now seems it’s not that constructive. In particular, it does not work when the user uses no semicolons.

@michaelficarra, @nzakas, would you mind giving me a short crash-course and a few directions on how do I use the escope information?

**Comment by nzakas** [comment]
This is the best place to start: http://estools.github.io/escope/

You can get Scope objects by using `context.getScope()` in your rule (use it inside a `Program` method to get the global scope). Then, you can use `childScopes` to recursively look through each other scope. Searching in this repo, you can find several rules that do similar things.

**Comment by denis-sokolov** [comment]
Thank you, @nzakas, I will investigate and consider a new PR.

**Pull request closed without merging**